endpoints/openrtb2/auction.go

```diff
@@ -335,28 +337,49 @@ func sendAuctionResponse(
 //
 // If the errors list has at least one element, then no guarantees are made about the returned request.
 func (deps *endpointDeps) parseRequest(httpRequest *http.Request, labels *metrics.Labels, hookExecutor hookexecution.HookStageExecutor) (req *openrtb_ext.RequestWrapper, impExtInfoMap map[string]exchange.ImpExtInfo, storedAuctionResponses stored_responses.ImpsWithBidResponses, storedBidResponses stored_responses.ImpBidderStoredResp, bidderImpReplaceImpId stored_responses.BidderImpReplaceImpID, account *config.Account, errs []error) {
-	req = &openrtb_ext.RequestWrapper{}
-	req.BidRequest = &openrtb2.BidRequest{}
 	errs = nil
-
-	// Pull the request body into a buffer, so we have it for later usage.
-	lr := &io.LimitedReader{
-		R: httpRequest.Body,
-		N: deps.cfg.MaxRequestSize,
+	var limitedReqReader *io.LimitedReader
+	reqContentEnc := httpRequest.Header.Get("Content-Encoding")
+	if reqContentEnc == "gzip" {
+		if !deps.cfg.Compression.Request.Enabled ||
+			deps.cfg.Compression.Request.CType != config.CompressionTypeGZIP {
+			errs = []error{fmt.Errorf("Content-Encoding of type %s is not supported", reqContentEnc)}
+			return
+		}
+		if gzReader, err := gzip.NewReader(httpRequest.Body); err == nil {
+			limitedReqReader = &io.LimitedReader{
+				R: gzReader,
+				N: deps.cfg.MaxRequestSize,
+			}
+			defer gzReader.Close()
+		} else {
+			errs = []error{err}
+			return
+		}
+	} else {
+		limitedReqReader = &io.LimitedReader{
+			R: httpRequest.Body,
+			N: deps.cfg.MaxRequestSize,
+		}
 	}
-	requestJson, err := io.ReadAll(lr)
+
+	requestJson, err := ioutil.ReadAll(limitedReqReader)
 	if err != nil {
 		errs = []error{err}
 		return
 	}
-	// If the request size was too large, read through the rest of the request body so that the connection can be reused.
-	if lr.N <= 0 {
-		if written, err := io.Copy(io.Discard, httpRequest.Body); written > 0 || err != nil {
```

Comment:
This was updated to the fillowing logic because in case of gzipped request this logic doesn't work as expected. Reason being, we first decompress the entire request and then apply a `limitedReader` on the decompressed request and therefore we won't have any leftover data in the `httpRequest.Body` even if the request was over our `MaxRequestSize` limit.

The below logic doesn't depend on the `httpRequest.Body` but rather checks if there's more to read in the `LimitedReader`

Reply:
I'm concerned having code to read through the entirely of a compressed file exposes PBS to "zip bombs". Do we need to check before calling `io.Copy(io.Discard, httpRequest.Body)`? Would it always be a good idea to ensure the request body is fully read to avoid the connection reuse limitation? The code is old, is the connection reuse issue still relevant for Go 1.19+?

Reply:
> I'm concerned having code to read through the entirely of a compressed file exposes PBS to "zip bombs"

The current request size limit that we have is on an uncompressed request. We will spend resources decompressing a large request in case of a zip bomb but won't be spending resources processing the request since we're using limited reader. If we want to avoid spending resources even decompressing the request, we can add another limit specific for compressed requests which is much lower than the uncompressed request limit. However, choosing this value might be a little tricky as different compression methods and their different compression levels might lead to different compression ratios. This is not to say we shouldn't add it.

> Do we need to check before calling io.Copy(io.Discard, httpRequest.Body)? Would it always be a good idea to ensure the request body is fully read to avoid the connection reuse limitation? The code is old, is the connection reuse issue still relevant for Go 1.19+?

So, looks like the http spec recommends that a server returns 413 and closes the connection in case of a request that's larger than what it allows. However, go doesn't do this by default. I assumed that it did because of the comment here but found otherwise after I tested it using a client. So we can just get rid of the `io.Copy(io.Discard, httpRequest.Body)` if we want to but it doesn't really hurt to have it there. In regards to the check, it is still needed to return an error if the request is too large.